# To do:

- build dataset
- test `AcousticDatafile(test.wav)` and `segments` method
- test `AcousticDataset(directory)` and `segments` method

In [1]:
from dnnseg.data import AcousticDatafile, AcousticDataset

In [2]:
import os

In [43]:
os.chdir("D:/model/german/")

In [44]:
file_names = [x[:-4] for x in os.listdir() if x.endswith(".wav")][:7]

In [45]:
file_names

['de2_e_b01_v0102301011111',
 'de2_e_b01_v0102301011112',
 'de2_e_b01_v0102301011113',
 'de2_e_b01_v0102301011114',
 'de2_e_b01_v0102301011115',
 'de2_e_b01_v0102301011116',
 'de2_e_b01_v0102301011117']

In [46]:
dataset = AcousticDataset("D:/tests/")

Processing file "de2_e_b..." 7/7       |    ETA - 0s     


Solving issue in .inputs and .targets methods inside AcousticDataset class.
Notes: `.inputs` and `.targets` only call `.segment_and_stack`, once as `padding = 'pre'` and once `padding = 'post'`
# To do:
- test `.segment_and_stack`from `AcousticDataset` outside of script 
- test `.inputs` and `.targets` outside of script
- UNDERSTAND what this does

In [9]:
dataset.targets()

using method .targets from AcousticDataset

 ####### using method .segment_and_stack from AcousticDataset ###### 

ARGUMENTS PASSED TO METHOD SEGMENT_AND_STACK IN ACOUSTICDATASET 

segments = wrd 

max_len = None 

reverse = True 

normalize = False 

center = False 

with_deltas = False 

resample = None 

ARGUMENTS PASSED TO METHOD SEGMENT_AND_STACK IN ACOUSTICDATAFILE: 

segments = wrd 

max_len = None 

padding = post 

reverse = True 

normalize = False 

center = False 

with_deltas = False 

resample = None 



(array([], shape=(8, 0, 13), dtype=float64),
 array([], shape=(8, 0), dtype=float64))

In [10]:
dataset.segment_and_stack()


 ####### using method .segment_and_stack from AcousticDataset ###### 

ARGUMENTS PASSED TO METHOD SEGMENT_AND_STACK IN ACOUSTICDATASET 

segments = wrd 

max_len = None 

reverse = False 

normalize = False 

center = False 

with_deltas = True 

resample = None 

ARGUMENTS PASSED TO METHOD SEGMENT_AND_STACK IN ACOUSTICDATAFILE: 

segments = wrd 

max_len = None 

padding = pre 

reverse = False 

normalize = False 

center = False 

with_deltas = True 

resample = None 



ValueError: all the input array dimensions for the concatenation axis must match exactly, but along dimension 1, the array at index 0 has size 43510 and the array at index 1 has size 56370

In [11]:
dataset.inputs()


 ####### using method .inputs from AcousticDataset ###### 


 ####### using method .segment_and_stack from AcousticDataset ###### 

ARGUMENTS PASSED TO METHOD SEGMENT_AND_STACK IN ACOUSTICDATASET 

segments = wrd 

max_len = None 

reverse = False 

normalize = False 

center = False 

with_deltas = True 

resample = None 

ARGUMENTS PASSED TO METHOD SEGMENT_AND_STACK IN ACOUSTICDATAFILE: 

segments = wrd 

max_len = None 

padding = pre 

reverse = False 

normalize = False 

center = False 

with_deltas = True 

resample = None 



ValueError: all the input array dimensions for the concatenation axis must match exactly, but along dimension 1, the array at index 0 has size 43510 and the array at index 1 has size 56370

In [50]:
datafile = dataset.data["de2_e_b01_v0102301011114"]

In [9]:
datafile.segment_and_stack(padding="pre")

(array([[[-4.3935999e+02,  3.8676880e+01,  4.1746397e+00, ...,
          -1.3803151e+00, -7.7529901e-01,  1.0259549e-01],
         [-4.3328519e+02,  5.8681412e+01,  1.4759262e+01, ...,
          -1.0292969e+00, -1.0466719e+00,  4.4414034e-01],
         [-4.0103995e+02,  8.3274643e+01,  2.6987366e+01, ...,
           6.4401424e-01,  8.8818264e-01,  5.4182857e-01],
         ...,
         [-3.7044510e+02,  1.0567401e+02,  3.3955765e+01, ...,
           1.0104920e-01, -1.5401207e-01,  5.6449831e-01],
         [-3.7411316e+02,  9.9037827e+01,  3.1980671e+01, ...,
          -5.0408220e-01,  1.0673784e-01,  9.2395109e-01],
         [-3.8897562e+02,  9.1623596e+01,  3.5974098e+01, ...,
          -9.9924690e-01,  3.4237463e-02,  3.6131066e-01]]], dtype=float32),
 array([[1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
         1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
         1., 1., 1., 1., 1., 1., 1., 1., 1., 1.]]))

In [10]:
datafile.segment_and_stack(padding="post")

(array([[[-4.3935999e+02,  3.8676880e+01,  4.1746397e+00, ...,
          -1.3803151e+00, -7.7529901e-01,  1.0259549e-01],
         [-4.3328519e+02,  5.8681412e+01,  1.4759262e+01, ...,
          -1.0292969e+00, -1.0466719e+00,  4.4414034e-01],
         [-4.0103995e+02,  8.3274643e+01,  2.6987366e+01, ...,
           6.4401424e-01,  8.8818264e-01,  5.4182857e-01],
         ...,
         [-3.7044510e+02,  1.0567401e+02,  3.3955765e+01, ...,
           1.0104920e-01, -1.5401207e-01,  5.6449831e-01],
         [-3.7411316e+02,  9.9037827e+01,  3.1980671e+01, ...,
          -5.0408220e-01,  1.0673784e-01,  9.2395109e-01],
         [-3.8897562e+02,  9.1623596e+01,  3.5974098e+01, ...,
          -9.9924690e-01,  3.4237463e-02,  3.6131066e-01]]], dtype=float32),
 array([[1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
         1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
         1., 1., 1., 1., 1., 1., 1., 1., 1., 1.]]))

`segment_and_stack` method from AcousticDatafile

Arguments passed during fitting:
```
segments = "wrd",
max_len = 50,
padding = "pre",
reverse = False,
normalize = False,
center = True,
with_deltas = True,
resample = None
```

In [14]:
segments_arr = datafile.wrd_segments

In [15]:
segments_arr

fileID  start      end speaker label
0  de2_e_b01_v0102301011114  524.4  939.798     v01   zUN

In [22]:
padding

'pre'

In [20]:
pad_seqs = padding not in ['None', None]

In [21]:
pad_seqs

True

In [11]:
from dnnseg.data import pad_sequence

In [12]:
import numpy as np

In [16]:
bounds = np.array(np.rint((segments_arr[['start', 'end']] / datafile.offset)),
                  dtype=np.int32)

In [17]:
bounds

array([[52, 94]])

`bounds` were the boundaries in which there is voice in ms * 1000 / 10 (because the offset is 10), now corrected

In [64]:
bounds.shape

(1, 2)

In [65]:
datafile.offset

10

In [18]:
import scipy

In [24]:
segments = "wrd"
max_len = 1300
padding = "pre"
reverse = False
normalize = False
center = True
with_deltas = True
resample = None

feats_split = []
mask_split = []

print(f"feats = {datafile.feats}")
print(f"feats shape is {datafile.feats.shape}")

for i in range(len(bounds)):
    print(f"loop no. {i}")
    s = bounds[i, 0]
    e = bounds[i, 1]
    print(f"s = {s}")
    print(f"e = {e}")
    
    if max_len is not None:
        print("cutting to maximum length")
        # cut features to maximum length
        e = min(e, s + max_len)
    new_feats = datafile.feats[s:e, :]
    length = e - s
    print(f"new_feats = {new_feats}")
    print(f"new_feats shape is {new_feats.shape}")
    print(f"length = {length}")
    
    
    if resample and new_feats.shape[0] > 0:
        print("resampling")
        length = resample
        new_feats = scipy.signal.resample(new_feats, resample, axis=0)
        print(f"new_feats = {new_feats}")

    if not with_deltas:
        print("cutting deltas")
        new_feats = new_feats[:, :datafile.n_coef]
        print(f"new_feats = {new_feats}")

    if normalize:
        print("normalizing")
        maximum = new_feats.max()
        minimum = new_feats.min()
        diff = maximum - minimum
        new_feats = (new_feats - minimum) / diff
        print(f"new_feats = {new_feats}")
    if center:
        print("before centering")
        print(f"new_feats = {new_feats}")
        print(f"new_feats.mean = {new_feats.mean()}")
        print(f"Are feats and mean the same? --> {new_feats == new_feats.mean()}")
        new_feats = new_feats - new_feats.mean()
        print("after centering")
        print(f"new_feats = {new_feats}")

    if not pad_seqs:
        print("pad_seqs is None")
        new_feats = new_feats[None, ...]
        mask = np.ones((1, length))
        print(f"new_feats = {new_feats}")
    else:
        print(f"creating mask with length of {length}")
        mask = np.ones(length)
        print(f"new_feats = {new_feats}")

    feats_split.append(new_feats)
    mask_split.append(mask)
    
    print("feats_split:")
    print(feats_split)
    print("\n")
    print("feats length:")
    print(len(feats_split))
    print("\n")
    print("mask_split:")
    print(mask_split)
    print("\n")
    print("mask length:")
    print(len(mask_split))
    

if pad_seqs:
    feats = pad_sequence(feats_split, padding=padding, reverse=reverse)
    mask = pad_sequence(mask_split, padding=padding, reverse=reverse)
else:
    feats = feats_split
    mask = mask_split

feats = [[-4.1554306e+02  6.9378067e+01  1.6545097e+01 ...  2.3677826e-01
  -1.1491722e+00 -1.0111792e+00]
 [-4.1187183e+02  5.6095604e+01 -2.0172679e-01 ...  2.3677826e-01
  -1.1491722e+00 -1.0111792e+00]
 [-4.1587122e+02  6.0025391e+01  9.4111891e+00 ...  2.3677826e-01
  -1.1491722e+00 -1.0111792e+00]
 ...
 [-4.1675873e+02  5.5765411e+01  3.2940605e+00 ... -2.2276479e-01
  -6.1275685e-01 -8.8081872e-01]
 [-4.2713913e+02  4.2217934e+01 -3.4910685e-01 ... -2.2276479e-01
  -6.1275685e-01 -8.8081872e-01]
 [-4.2872861e+02  5.1364624e+01  1.3756428e+01 ... -2.2276479e-01
  -6.1275685e-01 -8.8081872e-01]]
feats shape is (149, 39)
loop no. 0
s = 52
e = 94
cutting to maximum length
new_feats = [[-4.3935999e+02  3.8676880e+01  4.1746397e+00 ... -1.3803151e+00
  -7.7529901e-01  1.0259549e-01]
 [-4.3328519e+02  5.8681412e+01  1.4759262e+01 ... -1.0292969e+00
  -1.0466719e+00  4.4414034e-01]
 [-4.0103995e+02  8.3274643e+01  2.6987366e+01 ...  6.4401424e-01
   8.8818264e-01  5.4182857e-01]
 ...
 [

In [98]:
datafile.feats[500:502]

array([], shape=(0, 39), dtype=float32)

In [50]:
datafile.feats.shape

(149, 39)

In [52]:
datafile.n_coef

13

In [68]:
type(s)

numpy.int32

In [74]:
new_feats = datafile.feats - datafile.feats.mean()

In [76]:
new_feats.shape

(149, 39)

## error causes found:

1) `max_len` needs to be adapted to this material in config file. Originally it was 50, which yields an empty array at `segment_and_stack`. It was supposed to be the maximum syllable length in the material set.

2) `self.bounds`, the boundaries in which there is voice in ms * 1000 / offset (offset is defined by `self.offset`), are yielding a slice of `self.features` that is empty. `self.features.shape` is `(149, 39)`, the boundaries are being set at over 5000 

## to do:
- find maximum syllable length in material set (`D:/MASTER/23-LIN-MaCL1/Modulprüfung/dfc3sd.csv`)
- set found length (in ms) in config file `max_len`
- correct boundaries method to yield boundaries under 149 

In [25]:
feats.shape

(1, 42, 39)

In [26]:
mask.shape

(1, 42)

In [9]:
dataset.labels()

array([[1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.]])

In [11]:
from dnnseg.backend import binary2integer

In [13]:
import tensorflow as tf

In [16]:
import numpy as np

In [14]:
test = tf.Session()

In [20]:
b2i = binary2integer(b=np.array([[True, True, False],[True, True, False]]), session=test)

In [24]:
b2i

<tf.Tensor 'Sum_3:0' shape=(2,) dtype=int32>

In [21]:
with test:
    b2i.eval()

In [22]:
b2i.value_index

0

In [23]:
np.array([[True, True, False],[True, True, False]]).shape

(2, 3)

In [26]:
x, x_mask = dataset.inputs()


 ####### using method .inputs from AcousticDataset ###### 


 ####### using method .segment_and_stack from AcousticDataset ###### 

ARGUMENTS PASSED TO METHOD SEGMENT_AND_STACK IN ACOUSTICDATASET 

segments = wrd 

max_len = None 

reverse = False 

normalize = False 

center = False 

with_deltas = True 

resample = None 

ARGUMENTS PASSED TO METHOD SEGMENT_AND_STACK IN ACOUSTICDATAFILE: 

segments = wrd 

max_len = None 

padding = pre 

reverse = False 

normalize = False 

center = False 

with_deltas = True 

resample = None 



In [35]:
indices = np.arange(0, min(0 + 256, len(dataset.targets()[1])))

using method .targets from AcousticDataset

 ####### using method .segment_and_stack from AcousticDataset ###### 

ARGUMENTS PASSED TO METHOD SEGMENT_AND_STACK IN ACOUSTICDATASET 

segments = wrd 

max_len = None 

reverse = True 

normalize = False 

center = False 

with_deltas = False 

resample = None 

ARGUMENTS PASSED TO METHOD SEGMENT_AND_STACK IN ACOUSTICDATAFILE: 

segments = wrd 

max_len = None 

padding = post 

reverse = True 

normalize = False 

center = False 

with_deltas = False 

resample = None 



In [36]:
indices

array([0, 1, 2, 3, 4, 5, 6, 7])

In [27]:
x.shape

(8, 56, 39)

In [33]:
x[indices]

array([[[ 0.00000000e+00,  0.00000000e+00,  0.00000000e+00, ...,
          0.00000000e+00,  0.00000000e+00,  0.00000000e+00],
        [ 0.00000000e+00,  0.00000000e+00,  0.00000000e+00, ...,
          0.00000000e+00,  0.00000000e+00,  0.00000000e+00],
        [ 0.00000000e+00,  0.00000000e+00,  0.00000000e+00, ...,
          0.00000000e+00,  0.00000000e+00,  0.00000000e+00],
        ...,
        [-3.90040466e+02,  8.03762512e+01,  1.63609600e+01, ...,
          9.30139124e-01,  1.36380458e+00,  2.19446969e+00],
        [-4.02878571e+02,  7.99380264e+01,  2.69165039e+01, ...,
          2.31215739e+00,  2.40860820e+00,  1.80985677e+00],
        [-4.19096375e+02,  6.18056870e+01,  2.19476452e+01, ...,
          1.34008563e+00,  7.01640368e-01, -7.13101268e-01]],

       [[-4.15527161e+02,  3.69979858e+01,  6.25174046e+00, ...,
          1.53874552e+00,  2.88899809e-01, -3.28236490e-01],
        [-4.22955078e+02,  3.70714722e+01,  1.10998583e+01, ...,
          5.60496688e-01,  8.97203743e

In [34]:
x_mask[indices]

array([[0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 1., 1., 1.,
        1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
        1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
        1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
        1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
        1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
        1., 1., 1., 1., 1., 1., 1., 1.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
        1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
        1., 1., 1., 1., 1., 1., 1., 1.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 1.,
        1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
        1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
        1., 1

In [47]:
x, x_mask = dataset.segment_and_stack()


 ####### using method .segment_and_stack from AcousticDataset ###### 

ARGUMENTS PASSED TO METHOD SEGMENT_AND_STACK IN ACOUSTICDATASET 

segments = wrd 

max_len = None 

reverse = False 

normalize = False 

center = False 

with_deltas = True 

resample = None 

ARGUMENTS PASSED TO METHOD SEGMENT_AND_STACK IN ACOUSTICDATAFILE: 

segments = wrd 

max_len = None 

padding = pre 

reverse = False 

normalize = False 

center = False 

with_deltas = True 

resample = None 



In [48]:
x.shape

(7, 56, 39)

In [49]:
dataset.data

{'de2_e_b01_v0102301011111': <dnnseg.data.AcousticDatafile at 0x1ea60550518>,
 'de2_e_b01_v0102301011112': <dnnseg.data.AcousticDatafile at 0x1ea60550438>,
 'de2_e_b01_v0102301011113': <dnnseg.data.AcousticDatafile at 0x1ea60550160>,
 'de2_e_b01_v0102301011114': <dnnseg.data.AcousticDatafile at 0x1ea60550978>,
 'de2_e_b01_v0102301011115': <dnnseg.data.AcousticDatafile at 0x1ea604d8710>,
 'de2_e_b01_v0102301011116': <dnnseg.data.AcousticDatafile at 0x1ea604aa1d0>,
 'de2_e_b01_v0102301011117': <dnnseg.data.AcousticDatafile at 0x1ea5f34ecf8>}

In [54]:
x, x_mask = datafile.segment_and_stack(padding="pre", max_len=30)

In [55]:
x.shape

(1, 30, 39)

In [56]:
x, x_mask = dataset.segment_and_stack(max_len=30)


 ####### using method .segment_and_stack from AcousticDataset ###### 

ARGUMENTS PASSED TO METHOD SEGMENT_AND_STACK IN ACOUSTICDATASET 

segments = wrd 

max_len = 30 

reverse = False 

normalize = False 

center = False 

with_deltas = True 

resample = None 

ARGUMENTS PASSED TO METHOD SEGMENT_AND_STACK IN ACOUSTICDATAFILE: 

segments = wrd 

max_len = 30 

padding = pre 

reverse = False 

normalize = False 

center = False 

with_deltas = True 

resample = None 



In [57]:
x.shape

(7, 30, 39)